### PR TITLE
[Moore] SymbolVisibility attribute support for SVModuleOp

### PR DESCRIPTION
**Pull request by cepheus69**
We can append the appropriate symbol visibility to svmoduleOp according to the structures of instanceSymbol provided by the slang frontend. slang provides function to get the root of the design. Calling this method could get all top-level instanceSymbols and help determine which ones should be tagged. Note that the visibility attribute now used does not contain the `nested`.

### CODE REVIEW

**Comment by cepheus69** [comment]
> Can we define visibility in ODS like https://github.com/llvm/llvm-project/blob/56f0ecd6db9219b7d14a8eda613d6b75060643eb/mlir/include/mlir/Dialect/Func/IR/FuncOps.td#L271? I know many CIRCT ops are not following this convention but probably it's better to use this style. We should also fix other CIRCT ops.

Well, it is a good choice to follow the upstream style. I'll take a look.

File: lib/Dialect/Moore/MooreOps.cpp [hunk 1]
```diff
@@ -25,19 +25,38 @@ using namespace mlir;
...
                                                 getModuleType(), {}, {});
-  p.printOptionalAttrDictWithKeyword(getOperation()->getAttrs(),
-                                     getAttributeNames());
   p << " ";
   p.printRegion(getBodyRegion(), /*printEntryBlockArgs=*/false,
                 /*printBlockTerminators=*/true);
+
+  SmallVector<StringRef, 4> omittedAttrs = {"sym_name", "module_type",
+                                            visibilityAttrName};
+  p.printOptionalAttrDictWithKeyword((*this)->getAttrs(), omittedAttrs);
```
Comment (uenoku) [inline lib/Dialect/Moore/MooreOps.cpp:53]: I think this change is not required anymore? Also this changes the position of attributes and is that intentional?

Comment (cepheus69) [reply lib/Dialect/Moore/MooreOps.cpp:53]: You are right. It is no need to preserve it.

Comment (cepheus69) [reply lib/Dialect/Moore/MooreOps.cpp:53]: I will delete it soon. Thanks so much.

File: lib/Dialect/Moore/MooreOps.cpp [hunk 2]
```diff
@@ -25,19 +25,34 @@ using namespace mlir;
...
+  // Print the visibility of the module.
+  StringRef visibilityAttrName = SymbolTable::getVisibilityAttrName();
+  if (auto visibility = (*this)->getAttrOfType<StringAttr>(visibilityAttrName))
+    p << visibility.getValue() << ' ';
+
   p.printSymbolName(SymbolTable::getSymbolName(*this).getValue());
   hw::module_like_impl::printModuleSignatureNew(p, getBodyRegion(),
                                                 getModuleType(), {}, {});
-  p.printOptionalAttrDictWithKeyword(getOperation()->getAttrs(),
-                                     getAttributeNames());
```
Comment (uenoku) [inline lib/Dialect/Moore/MooreOps.cpp:34]: We should keep this.

**Review by uenoku — APPROVED** [review]
Please revert the change regarding `printOptionalAttrDictWithKeyword` otherwise LGTM

**Comment by cepheus69** [comment]
Thank you for the review!

File: lib/Dialect/Moore/MooreOps.cpp (hunk 2 above)
Comment (cepheus69) [reply lib/Dialect/Moore/MooreOps.cpp:34]: I think I misunderstood your previous comments. It was already been reverted and applied the new way.

**Pull request merged**